packages/vite-plugin-cloudflare/src/modules-plugin.ts

```diff
@@ -0,0 +1,102 @@
+import assert from "node:assert";
+import * as fs from "node:fs";
+import * as path from "node:path";
+import MagicString from "magic-string";
+import * as vite from "vite";
+import { MODULE_PATTERN } from "./shared";
+import type { ResolvedPluginConfigContainer } from ".";
+import type { ModuleType } from "./constants";
+
+function createModuleReference(type: ModuleType, id: string) {
+	return `__CLOUDFLARE_MODULE__${type}__${id}__`;
+}
+
+export function modulesPlugin(
+	resolvedPluginConfigContainer: ResolvedPluginConfigContainer
+): vite.Plugin {
+	return {
+		name: "vite-plugin-cloudflare:modules",
+		// We set `enforce: "pre"` so that this plugin runs before the Vite core plugins.
+		// Otherwise the `.wasm` extension cannot be used for module imports
+		enforce: "pre",
+		applyToEnvironment(environment) {
+			if (
+				resolvedPluginConfigContainer.resolvedPluginConfig.type ===
+				"assets-only"
+			) {
+				return false;
+			}
+
+			return Object.keys(
+				resolvedPluginConfigContainer.resolvedPluginConfig.workers
+			).includes(environment.name);
+		},
+		async resolveId(source, importer) {
+			if (!source.endsWith(".wasm")) {
+				return;
+			}
+
+			const resolved = await this.resolve(source, importer);
+			assert(
+				resolved,
+				`Unexpected error: could not resolve Wasm module ${source}`
+			);
+
+			return {
+				external: true,
+				id: createModuleReference("CompiledWasm", resolved.id),
+			};
+		},
+		renderChunk(code, chunk) {
+			const moduleRE = new RegExp(MODULE_PATTERN, "g");
+			let match: RegExpExecArray | null;
+			let s: MagicString | undefined;
```

Comment:
nit: I'm usually against single letter variables, especially if they are used in more than a handful of lines of code, so here it would be nice if we could have a better name for the variable, although yeah I do think that this is a bit awkward and I can come up with a great name either 😕 🤷 

`codeUpdater`? `codeMagicString`? `content`?

it's up to you, I'm also fine with `s` 👍

Reply:
Yeah, totally agree about single letter variables but couldn't think of a good name. I'll just call it `magicString`.